File: ios/Classes/PayvizioPlugin.swift
import Flutter
import UIKit
import Payvizio

/**
 MethodChannel bridge from Dart to the Payvizio iOS SDK. Resolves the
 top-most view controller at checkout-time so the hosted checkout sheet has a
 presenter regardless of how the host app's navigation stack is built.
 */
public class PayvizioPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.payvizio.flutter/sdk",
            binaryMessenger: registrar.messenger())
        let instance = PayvizioPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":       handleConfigure(call, result: result)
        case "prefetch":        Payvizio.shared.prefetch(); result(nil)
        case "checkout":        handleCheckout(call, result: result)
        case "launchUpiIntent": handleUpi(call, result: result)
        default:                result(FlutterMethodNotImplemented)
        }
    }

    private func handleConfigure(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let baseUrlString = args["apiBaseUrl"] as? String,
              let baseUrl = URL(string: baseUrlString) else {
            result(FlutterError(code: "PVZ_INVALID", message: "apiBaseUrl is required", details: nil)); return
        }
        let checkoutUrl = (args["checkoutUrl"] as? String).flatMap(URL.init(string:))
        let pollMs = (args["pollIntervalMs"] as? Int) ?? 2500
        let dismissible = (args["dismissible"] as? Bool) ?? true
        Payvizio.shared.configure(PayvizioConfig(
            apiBaseUrl: baseUrl,
            checkoutUrl: checkoutUrl,
            pollInterval: TimeInterval(pollMs) / 1000.0,
            dismissible: dismissible))
        result(nil)
    }

    @MainActor
    private func handleCheckout(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sessionId = args["sessionId"] as? String, !sessionId.isEmpty else {
            result(FlutterError(code: "PVZ_INVALID", message: "sessionId is required", details: nil)); return
        }
        guard let presenter = topViewController() else {
            result(FlutterError(code: "PVZ_NO_VC", message: "No top view controller", details: nil)); return
        }
        let cb = BridgeCallback(sessionId: sessionId, result: result)
        _ = Payvizio.shared.checkout(presenting: presenter, sessionId: sessionId, callback: cb)
    }

    @MainActor
    private func handleUpi(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let url = args["url"] as? String else {
            result(FlutterError(code: "PVZ_INVALID", message: "url is required", details: nil)); return
        }
        UpiIntent.launch(url) { ok in result(ok) }
    }

    @MainActor
    private func topViewController() -> UIViewController? {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        let window = scenes.flatMap { $0.windows }.first { $0.isKeyWindow }
        var top = window?.rootViewController
        while let presented = top?.presentedViewController { top = presented }
        return top
    }
}

/// Boxes the Flutter result so the lifecycle-bound callback can hand back a
/// single map and translate close-without-terminal into a synthetic CANCELLED.
private final class BridgeCallback: PaymentCallback {

    let sessionId: String
    var result: FlutterResult?

    init(sessionId: String, result: @escaping FlutterResult) {
        self.sessionId = sessionId
        self.result = result
    }

    func onSuccess(_ r: PaymentResult) { deliver(r) }
    func onFailure(_ r: PaymentResult) { deliver(r) }
    func onClose() {
        guard let r = result else { return }
        r(["sessionId": sessionId, "status": "CANCELLED"])
        result = nil
    }

    private func deliver(_ r: PaymentResult) {
        guard let cb = result else { return }
        cb([
            "sessionId":        r.sessionId,
            "status":           r.status.rawValue,
            "acquirer":         r.acquirer ?? NSNull(),
            "gatewayReference": r.gatewayReference ?? NSNull(),
            "amount":           r.amount ?? NSNull(),
            "currency":         r.currency ?? NSNull(),
            "failureReason":    r.failureReason ?? NSNull(),
        ])
        result = nil
    }
}

File: android/src/main/kotlin/com/payvizio/flutter/PayvizioPlugin.kt
package com.payvizio.flutter

import android.app.Activity
import com.payvizio.sdk.PaymentCallback
import com.payvizio.sdk.PaymentResult
import com.payvizio.sdk.Payvizio
import com.payvizio.sdk.PayvizioConfig
import com.payvizio.sdk.UpiIntent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * MethodChannel bridge from Dart to the Payvizio Android SDK. Holds the
 * current Activity binding so [com.payvizio.sdk.Payvizio.checkout] has a
 * launcher to open CheckoutActivity from.
 */
class PayvizioPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.payvizio.flutter/sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivity()                              { activity = null }
    override fun onReattachedToActivityForConfigChanges(b: ActivityPluginBinding) { activity = b.activity }
    override fun onDetachedFromActivityForConfigChanges()              { activity = null }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "configure"        -> handleConfigure(call, result)
            "prefetch"         -> { Payvizio.prefetch(); result.success(null) }
            "checkout"         -> handleCheckout(call, result)
            "launchUpiIntent"  -> handleUpiIntent(call, result)
            else               -> result.notImplemented()
        }
    }

    private fun handleConfigure(call: MethodCall, result: MethodChannel.Result) {
        val baseUrl = call.argument<String>("apiBaseUrl") ?: run {
            result.error("PVZ_INVALID", "apiBaseUrl is required", null); return
        }
        val checkoutUrl = call.argument<String?>("checkoutUrl")
        val pollMs = (call.argument<Number>("pollIntervalMs") ?: 2500).toLong()
        val dismissible = call.argument<Boolean>("dismissible") ?: true
        Payvizio.init(PayvizioConfig(
            apiBaseUrl = baseUrl,
            checkoutUrl = checkoutUrl,
            pollIntervalMs = pollMs,
            backButtonDismissible = dismissible))
        result.success(null)
    }

    private fun handleCheckout(call: MethodCall, result: MethodChannel.Result) {
        val sessionId = call.argument<String>("sessionId") ?: run {
            result.error("PVZ_INVALID", "sessionId is required", null); return
        }
        val act = activity ?: run {
            result.error("PVZ_NO_ACTIVITY", "No foreground Activity", null); return
        }
        Payvizio.checkout(act, sessionId, object : PaymentCallback {
            override fun onSuccess(r: PaymentResult) { result.success(toMap(r)) }
            override fun onFailure(r: PaymentResult) { result.success(toMap(r)) }
            override fun onClose() {
                // synthetic CANCELLED so Dart sees a single completion path
                result.success(mapOf("sessionId" to sessionId, "status" to "CANCELLED"))
            }
        })
    }

    private fun handleUpiIntent(call: MethodCall, result: MethodChannel.Result) {
        val url = call.argument<String>("url") ?: run {
            result.error("PVZ_INVALID", "url is required", null); return
        }
        val act = activity ?: run {
            result.error("PVZ_NO_ACTIVITY", "No foreground Activity", null); return
        }
        result.success(UpiIntent.launch(act, url))
    }

    private fun toMap(r: PaymentResult): Map<String, Any?> = mapOf(
        "sessionId"        to r.sessionId,
        "status"           to r.status.name,
        "acquirer"         to r.acquirer,
        "gatewayReference" to r.gatewayReference,
        "amount"           to r.amount,
        "currency"         to r.currency,
        "failureReason"    to r.failureReason,
    )
}
